### ISSUE TITLE
Data protocol

### ISSUE THREAD

**Issue by bilderbuchi** [pymeasure/experiment_control_protocol#3]
From https://github.com/pymeasure/pymeasure/discussions/676#discussioncomment-4285871

Benedikt:
Let us use this thread (in this discussion) to talk about the data transfer protocol, which seemed to be quite easy.

Here the current state of discussion:

- Data sources publish (PUB socket, connected to a sink or proxy) their data.
- Topic (first message frame) is a user defined name in order to subscribe only to those data sources you desire.
- Content (second message frame) is a dictionary serialized by json.
- Data sinks subscribe (SUB socket, bind to some port or connect to a proxy) to data sources.
- An optional proxy (XSUB-XPUB, bind to two different ports, based on `zmq.proxy_steerable`) allows several datasinks without changing the ports of the data sources.

Questions:
- which ports do we use as default? With a proxy, two ports are needed, one for data sources and one for data sinks.
- Do we want a proxy? I'd say yes, as it is little code (10-15 lines) but useful in some cases. It is easy to design the data sink as standalone capable (bind to the data source port instead of connect to data sink port) such that it is optional to use a proxy.
- Do we want to include the possibility to stream from one computer to several computers (e.g. localhost and some other)? The `zmq.proxy_steerable` can be easily adjusted (connecting to the "normal" proxy's ports on both computers) to connect two computers such that the data is streamed seamlessly (neither data source nor data sink take a notice, that they reside on different computers).
- Do we want an extra logging channel (I'd say yes, as it makes it easier to know where the log entry originated) or do we log with a specific topic (e.g. "log")?

pymeasure already implements a similar protocol in its `Worker` class:
- the worker creates a PUB socket and binds to a port (if desired).
- It publishes with different topics (error, results, status, progress). The content is pickled (with `cloudpickle`).
- The data published under "results" are dictionaries with the values.

With a few changes, we can make the pymeasure procedures publish values compatible to the proposed decentralized setup:
- connect to a port instead of binding.
- allow users to change the name from "results" to something different.
- Use the same serialization protocol as the other protocol does.
- Separate the data messages from logging (error, status, progress) and add the worker's name to the topics.

These changes, however, are breaking current setups at several points (we could, nevertheless, keep both paths open at the same time).

**Comment by bilderbuchi** [comment]
Benedikt:
Let us share the ZMQ protocols in order to understand, what the other does. All my zmq communication is in one [file](https://git.rwth-aachen.de/nloqo/eg-labor/-/blob/master/devices/devices/intercom.py), therefore, I can easily change that file and all the programs using it, won't notice anything.

### Protocol
- Data producers (measurement programs etc.) publish (PUB socket, connected to the proxy) a two frame message for every variable. This code is in the `Publisher` [class](https://git.rwth-aachen.de/nloqo/eg-labor/-/blob/master/devices/devices/intercom.py#L296). The first frame (which is the "topic") is the variable name (utf-8 encoded), the second frame contains the value pickled (the receiver supports json as well) for lossless communication.
- Controlling software uses a REQ port connecting to the software to control ROUTER port (e.g. DataLogger or a few other programs). This code is in the `Communicator` [class](https://git.rwth-aachen.de/nloqo/eg-labor/-/blob/master/devices/devices/intercom.py#L543). The message contains a single frame with a dictionary (pickled or encapsulated via json). If the dictionary's 'query' entry evaluates to True, the software has to send a response, otherwise a simple acknowledgement (`b"ACK"`) is sent. In case of an error (e.g. unpickling / unpacking json fails), `b"ERR" + errormessage` is sent.
- The data consumer (DataLogger or programs to control) can use the `Listener` [class](https://git.rwth-aachen.de/nloqo/eg-labor/-/blob/master/devices/devices/intercom.py#L357), which starts a background thread listening on incoming communication according to above protocol. You can control the communication (e.g. subscribing/unsubscribing) via methods of that class. If the class receives data (from subscribed to data producers of control commands), it emits PyQt Signals containing the data/command.


### Design considerations

Pro
- It is possible to subscribe to a single value of a data producer and you do not have to define the data producers to subscribe to **and** the variables to store (in the DataLogger).
- Pickling ensures that there is no data loss in comparison to json/converting to string, as that will cut float precision for example.
- To keep it simple, commands are not distinguished, but just a plain dictionary. The individual software has to determine what to do with specific dictionary entries.

Con
- Sending one message for every variable might put half of the variables of a data set (e.g. results of oscilloscope channels) in one data point and the other half in a second data point, if the data point generation happens in between. (Background: The DataLogger collects data in lists and, triggered by a timer or receiving a certain variable, generates one data point of the data received until then. E.g. if it receives 10 times some temperature variable and 1 voltage variable until triggered, the data point will contain the average, or last value, of the temperature and that voltage value)
- Reliance on pickle reduces interoperability with different programming languages (like Labview), for that reason I recently added json decoding.
- You need an individual port number for every program to control and you have to know, which program has which port.

**Comment by bilderbuchi** [comment]
Benjamin: 
Well, I guess pymeasure is, in part, about not inventing the wheel over and over again, so I would be extremely happy if we could bring these quite similar approaches together and find a good combination! Sadly I am not sure how much time I will be able to spend on it, I would have liked to do much more here on pymeasure, but this did not happen the last few months, so I am not sure it will happen in the next few months. 

Thanks a lot for your more detailed description of your approach, I think I can take the time to go a bit deeper into mine, I wanted to put that very detailed in my PhD thesis, but I can put it here first too: 

### Files
All the communication in my architecture is defined in [this folder](https://github.com/Cryostat-GUI/Cryostat-GUI/tree/master/CryostatGUI/util), specifically in the files starting with `broker**.py`, as well as `livedata.py` and `zmqcomms.py` 
I have not managed to pack everything in one file, and my implementation of continuously running processes which push out data all the time (but can be paused), found in the file `abstractThreads.py` comes with some less than readable things, like methods of class X being used in class Y "before" class Y has inherited from class X, so, way to go there in cleaning it up...

### Proxy/Broker/Channel structure

As you can see from the naming scheme, we too use proxies, in our case called brokers. 
We have 3 brokered channels:

1.  `downstream` (for commands to instruments, e.g. to change the temperature, `ip_maincontrol`), 
2. `upstream` (for data, continuously published `ip_data`) 
3. `tcp` (for a brokered req-repl pattern), 

and one unbrokered, direct communication channel `data`, more on that later. 
We have three principle types of programs, the base classes being defined in  `zmqcomms.py`: 

1. `zmqClient`, 

- controlling 1 specific instrument each, 
- publishing their data into `upstream`, 
- receiving commands via `downstream`, 
- answering questions on `tcp`
- multiple instances per measurement system allowed

2. `zmqDataStore`, 

- consumes data from `upstream`, 
  - saves them into lists of 60 entries, 
  - taking (I think/hope) the last consumed value for any variable once a second, 
  - by that building a list of the last minute, and performing simple calculations like the arithmetic mean, derivative and so on
- makes all its data available on the unbrokered channel `data`, which can be requested by a main Control, 
- one single instance per measurement system allowed (`data` channel is not brokered, could be changed though)

3. `zmqMainControl`, 

- sending commands via `downstream` 
- reading data mostly via the `data` channel from a `zmqDataStore`, 
- also directly inquire for data (and also send commands with the inquiry) through the `tcp` channel. 
- multiple instances per measurement system allowed

### zmq details

All zmq messages have two frames, where the first one is a device `identity`, typically either its own (for publishing data into `upstream` by `ControlClients`) or the one where it should go to (for commands in `downstream`, sent by a `zmqMainControl`), or the same in `tcp`, where the broker in `broker_reqp.py` sends such requests and their respective answers to and from such defined places. 
Data in messages is transmitted as json-packed dictionaries (so far I did not have problems with cut floating point numbers, we don't typically have such data). 
In the `tcp` channel, every message gets a unique message identifier `uuid`, so that unique answers can be sent to the respective question in a slightly asynchronous way (this solved quite a few issues, although I am not 100% sure we actually use any asynchronous patterns). 
Currently, `ControlClients` do not acknowledge purely command-like messages in `downstream`, although they do so in `tcp`. 

We use the first frame for the device `identity` instead of individual messages for individual datapoints, as we have multiple devices of exactly the same type (e.g. Keithley6221s), so that we would have to define this for every device with the respective identity or some other identifier on the device level. In the way we have it, we can change the usage of different devices on a higher level without having to change anything for individual device drivers or `ControlClient`s. 

The `zmqMainControl` has at least two implementations in our case:
1. A central GUI which will read all data and can also send control commands to the devices, being fully independent from the acting controller programs
2. A sequence running scheme, which can be 
- a very very simple script, in which the functions to send commands and request individual data from devices can be used to control all of them, similar to having one script which connects to the devices directly, but again quite strongly separated from the controlling processes, so that a failure in one of them does not interfere with anything else. 
- Something much more elaborate, found in [Sequence_zmq.py](https://github.com/Cryostat-GUI/Cryostat-GUI/blob/master/CryostatGUI/Sequence/Sequence_zmq.py), which was my try at recreating the convenience of a Quantum Design PPMS system, even going so far as to use (and reverse engineer) their MultiVu sequence generator, to use that instead of building my own (drag&drop and/or Copy&Paste are quite a pain to do for me as a physicist as main occupation, so I didn't). 
- typically this is what ramps the temperature, commands measurements to be taken, and saves data for individual runs into specified datafiles, separately from the logging parts of the `dataStorage`

The ports are fixed for the channels. 

### Prometheus+Grafana
in `livedata.py`, we have packed all the logic for sending all the data which goes into `upstream` to a prometheus database too. 
More on those can be found on their docs, [prometheus](https://prometheus.io/docs/introduction/overview/) and [grafana](https://grafana.com/grafana/). This combination is apparently being used by a variety of sysadmins to get quick overviews of all running systems, it is also used at a few synchrotron facilities within europe. 
If I had found a nice-and-easy way how to retrieve data from prometheus again with python, I might have totally forgone the `zmqDataStore` after finding out about prometheus....
For Prometheus, every instrument needs/gets a specific port to provide the data to be scraped by prometheus, defined in the code and a `yaml` for the prometheus configuration


### Conclusion
This is now somewhat of an overview, without graphical aid of figures, maybe I can find time to go over pro's and con's at a later date

**Comment by bilderbuchi** [comment]
Note: zeromq seems to be the obvious/ubiquitous candidate for the messaging.

Preliminary list of separate components
* instrument drivers
* continuously running data collectors/controllers/publishers
* GUI
* procedures
* API for datasinks to get all the data into preferred databases

**Comment by bilderbuchi** [comment]
Benjamin recounts details [here](https://github.com/pymeasure/pymeasure/discussions/676#discussioncomment-4252344) (quite long, I don't want to swampt the issue)

**Comment by bilderbuchi** [comment]
@bmoneke could you summarize the ZMQ/majordomo details discussed starting from [here](https://github.com/pymeasure/pymeasure/discussions/676#discussioncomment-4261170) -- it's too much to copy-paste as-is, I think.

**Comment by bmoneke** [comment]
First, thanks for all that organization, @bilderbuchi .

I thought we have two separate protocols: One only for downstreaming data, the _data protocol_ (e.g. based on PUB-SUB), and one for directed communication, the _control protocol_, where you can send commands to instruments etc.
Majordomo may inspire for the control protocol.

**Comment by bilderbuchi** [comment]
I'm not sure if those are so separable, wouldn't the publishing/subscribing also need to be handled/coordinated (e.g. via majordomo)? But yeah, we can still discuss them separately, I think granularity of topics will serve us well to keep discussion focused and finite. :-P

**Comment by bmoneke** [comment]
You do not need a coordinator: All Actors publish and the observer subscribes the actors, according to the required data. For a n-to-n connection we need a coordinator, which is an already written part of pyzmq ("proxy"): Actors connect to the coordinator's XSUB port, observers connect to the coordinator's XPUB port.
Done.

We only have to define the content of the published messages.

**Comment by bmoneke** [comment]
That data protocol can also be used for logging purposes.

There are several options to distinguish logging from measurement data:

- different ports (allows one observer for logging, another for measurement data)
- different topics (for example multimeter and log_multimeter) 
- different content
-

**Comment by bmoneke** [comment]
Just an example how simple a proxy (even with moving it to a thread and stopping it from the main thread) is:
~~~python
import zmq

def pub_sub_proxy(context):
    """Start a publisher subscriber proxy. Blocks until the context is terminated."""
    log.info("Start local proxy server.")
    s = context.socket(zmq.XSUB)
    p = context.socket(zmq.XPUB)
    s.bind("tcp://*:11100")
    p.bind("tcp://*:11099")
    try:
        zmq.proxy(p, s)
    except zmq.ContextTerminated:
        log.info("Proxy context terminated.")
    except Exception as exc:
        log.exception("Some other exception on proxy happened.", exc)

def start_proxy(context=None):
    """Start a proxy server in its own thread.

    .. code-block:: python

        context = start_proxy()
        # Stop the proxy:
        context.destroy()

    :param context: The zmq context. If None, it generates its own context.
    :return: The zmq context. To stop, call `context.destroy()`.
    """
    context = context or zmq.Context.instance()
    thread = threading.Thread(target=pub_sub_proxy, args=(context,))
    thread.daemon = True
    thread.start()
    log.info("Proxy thread started.")
    return context

if __name__ == "__main__":
    context = start_proxy()
    while True:
        if input("Quit with q:") == "q":
            context.destroy()
            break
~~~

**Comment by BenediktBurger** [comment]
What do you think about the name "event protocol" instead of "data protocol"?

Also the control protocol transmits data.
_Event_ describes a data protocol message quite good.

**Comment by bilderbuchi** [comment]
> What do you think about the name "event protocol" instead of "data protocol"?
> 
> Also the control protocol transmits data. _Event_ describes a data protocol message quite good.

I'm open to finding a better name. "Event protocol" sounds good, I can follow your reasoning. On the other hand, a control thing (e.g. a command) can also be understood as an "event", don't you agree?
Maybe we find more precise terms if we sharpen the distinction between what we want to happen in the "data" and "control" protocol variants, respectively?

Maybe we should add a concise and complete description of what we understand the eventually chosen terms to mean (e.g. in the glossary), so that that is clarified.
